props/blaster.h

```diff
@@ -159,59 +158,81 @@ class Blaster : public PROP_INHERIT_PREFIX PropBase {
     }
   }
 
-  bool auto_firing_ = false;
-  int shots_fired_ = 0;
-  bool is_jammed_ = false;
-
 #ifdef BLASTER_SHOTS_UNTIL_EMPTY
   const int max_shots_ = BLASTER_SHOTS_UNTIL_EMPTY;
 #else
   const int max_shots_ = -1;
 #endif
 
-  virtual int GetBulletCount() {
-    return max_shots_ - shots_fired_;
+  bool CheckEmpty() const {
+    return max_shots_ != -1 && shots_fired_ >= max_shots_;
+  }
+
+  virtual bool DoEmpty() {
+    if (CheckEmpty()) {
+      SaberBase::DoEffect(EFFECT_EMPTY, 0);  // Trigger the empty effect
+      return true;
+    }
+    return false;
   }
 
   virtual bool CheckJam(int percent) {
     int random = rand() % 100;
     return random < percent;
   }
 
-  virtual void Fire() {
+  bool DoJam() {
 #ifdef ENABLE_MOTION
 #ifdef BLASTER_JAM_PERCENTAGE
     // If we're already jammed then we don't need to recheck. If we're not jammed then check if we just jammed.
     is_jammed_ = is_jammed_ ? true : CheckJam(BLASTER_JAM_PERCENTAGE);
 
     if (is_jammed_) {
       SaberBase::DoEffect(EFFECT_JAM, 0);
-      return;
+      return true;
+    } else {
+      return false;
     }
 #endif
 #endif
+  }
 
-    if (max_shots_ != -1) {
-      if (shots_fired_ >= max_shots_) {
-        SaberBase::DoEffect(EFFECT_EMPTY, 0);
-        return;
-      }
-    }
+  virtual int GetBulletCount() {
```

Comment:
Seems the same as the inherited version, so we don't need it, right?

Reply:
I don't see anything overriding this. 
Did you mean we don't need the virtual?
I think that's correct.

Reply:
Sorry, I missed what file this was in.
It shows up as new because the diff is dumb. Maybe just move it up to line 172 to make the diff nicer.

Reply:
outdated